DDS-715 auth service affiliates

dmlond opened the pull request:
implements:
- get /auth_services/:id/affiliates
- get /auth_services/:id/affiliates/:uid
- post /auth_services/:id/affiliates/:uid/dds_user

adds concept of affiliate_search to auth_providers
auth_providers support affiliate_search if they belong_to an identity_provider

affiliate search endpoints return an exception if the auth_service does not support affiliate_search

dmlond committed:
adding net-ldap for ldap support

dmlond committed:
created identity_providers STI table
added identity_provider_id to authentication_services table for belongs_to relationship

dmlond committed:
AuthenticationService can belong_to an IdentityProvider

dmlond committed:
IdentityProvider
  - validates presence of host and port
LdapIdentityProvider
  - validates presence of ldap_base
  - implements affiliates(authentication_service, full_name_contains)
    - returns empty Array if full_name_contains is nil or less than 3 characters
    - performs an ldap query to the configured ldap host, port, and ldap_base
      filtering on displayName = *full_name_contains* and converts the returned
      entities into an array of unpersisted User objects attached to an unpersisted
      UserAuthenticationService object attached to the authentication_service provided

dmlond committed:
paired down the code for affiliates

dmlond committed:
moved 'fooby_entry' from before block to let variable, changed to 'expected_entry'

dmlond committed:
implemented affiliate(authentication_service, uid)
refactored to share ldap_search among both affiliates and affiliate

dmlond committed:
created an AuthenticationServicePreviewSerializer and spec

dmlond committed:
created AffiliateSerializer, which serializes a User object as an Affiliate

dmlond committed:
created shared_context 'mocked ldap' to standardize mocking of
the ldap search call in different specification tests.

ldap_identity_provider_spec uses the mocked ldap shared_context where appropriate

dmlond committed:
Authenticated Users can now create other users from ldap

dmlond committed:
added boolean flag for affiliate_search_supported

dmlond committed:
speling

dmlond committed:
AffiliateAuthenticationServiceSerializer is
similar to the AuthenticationServicePreviewSerializer, but
does not include the supports_identity_search field

dmlond committed:
no longer creates the authentication_service_provider object on the user

dmlond committed:
bug in affiliates endpoints will occur sporadically when
the username has a '.' or '@' sign, so modified to match
the way user_authentication_services factory works

dmlond committed:
No longer serializes the auth_provider

dmlond committed:
base.rb adds rescue handler for Net::LDAP::Error to return
proper 503 error with message that communication error occurred

ldap_shared_examples
  - modified mocked ldap to take returns and use it as is
  - implements an identity_provider dependant authentication_provider resource
    shared_example to test that the endpoint responds with the 503
    when a communication error occurs with the identity_provider,
    and a 400 occurs when the auth_provider does not support affiliate searches
  - an identity provider shared_example that sets up mocked ldap and
    expects the response to match the expected_response_status. Can wrap
    other expectations
  - an identified affiliate which asserts that the resource responds with
    a 404 when the uid does not match a user in the identity_provider
AuthProviderSpec expects AuthProvider to implement
- get /api/v1/auth_providers/:id/affiliates
- get /api/v1/auth_providers/:id/affiliates/:uid
- post /api/v1/auth_providers/:id/affiliates/:uid/dds_user

AuthProviderSpec meets expectation

dmlond committed:
- ldap_identity_provider affilates and affiliate methods
  - no longer require auth_provider argument
  - no longer create a user_authentication_service on the User

- auth_providers_api
  - auth_providers/:id/affiliates/:id
  - auth_providers/:id/affiliates/:id/dds_user
  - ensures that user_authenticaton_service is created with user on creation

dmlond committed:
Merge remote-tracking branch 'upstream/develop' into DDS-715_auth_service_affiliates

dmlond committed:
Merge pull request #901 from dmlond/DDS-715_auth_service_affiliates

DDS-715 auth service affiliates

dmlond commented:
I need to add rake tasks for identity_provider management. Hold off on the review

dmlond committed:
moved setting of ENV for auth_service creation to the
authentication_service:create shared_example

implemented rake tasks
  - auth_service:identity_provider:register
  - auth_services:identity_provider:remove
with tests.

dmlond committed:
implements and tests
  - rake identity_provider:ldap:create
  - rake identity_provider:destroy

dmlond committed:
Merge pull request #903 from dmlond/DDS-789_identity_provider_rake_tasks

DDS-789 identity provider rake tasks

dmann commented on app/api/dds/v1/auth_providers_api.rb:
```diff
@@ -28,6 +57,113 @@ class AuthProvidersAPI < Grape::API
+          if auth_service.identity_provider.nil?
+            unsupported_affiliate_search_error!
+          else
+            auth_service.identity_provider.affiliates(affiliate_params[:full_name_contains])
```
I think we're going to have to figure out a way to paginate these results at some point.

dmann replied:
Since this will be a breaking change, should we paginate before merging?

dmlond replied:
go ahead and merge. We will need to work through a cycle of customer feedback, Matt specifying API design for pagination, developer spiking LDAP pagination and implementation as a new feature.

dmann replied:
@Donmurry what do you think?

Donmurry replied:
Hmm, I think it might be best to add pagination in now, if possible.  If spike of LDAP pagination is going to take weeks vs days, then i say we can skip it.  But if we can incorporate it now, I think thats better than not.  We can weigh the cost of development now vs cost later.

dmann replied:
If the decision is to skip pagination, it would be good to inform the client developers of this decision and the reasoning behind it. I think they would appreciate a warning that, if they submit a query with 3 characters, they're likely to get a firehose of data OR at request timeout. That way when both of those outcomes eventually happen, hopefully, they won't make an "issue" out of it...

dmlond replied:
https://github.com/ruby-ldap/ruby-net-ldap/blob/e4c46a223a19feda78393a793711353aa1febdcd/lib/net/ldap/connection.rb

I really dont think the ruby net-ldap search interface allows you access to pagination. They paginate the results under the hood, but they repeatedly search until all results are returned before returning to the caller with all the results.  They do allow you to send a maximum size, and time value (this might be useful to prevent timeouts), but these just end up minimizing each call to get the next page's results, while still continuously getting each pages results until they are all returned.

If we want to support pagination, we will have to do it wastefully, e.g. get all results, skip to N, add until M, break. This would satisfy clients requirement if they end up needing pagination, but would not prevent server timeouts of memory explosions.

dmann replied:
I feel strongly that this endpoint needs to limit the number of returned results. The pagination default is 100, with a max of 1000. Ideally, the client would have the ability change this limit via the `per_page` parameter.

While it would be nice to do a full implementation of pagination, I don't believe that is necessary. That is, I don't think that we need to allow clients to use the `page` parameter. In fact, if `page` is provided, I suggest returning a "Not Implemented" or "Not supported by Identity Provider" error of some kind.

This is a technical limitation of ldap and that shouldn't dictate how our interface works. Let's say that you initially setup the affiliates endpoints with an identity provider that supported pagination and you were easily able to add support for `per_page` and `page`. Feature released, done. Now, you have to add a second identity provider that does not support pagination. What do you do? Remove pagination from the endpoint? Probably not. This is the way that I'm approaching this problem and hopefully provides context to my suggestions above.

dmann commented on app/models/identity_provider.rb:
```diff
@@ -0,0 +1,4 @@
+class IdentityProvider < ActiveRecord::Base
```
Since `IdentityProvider` is an interface, I think it would be good to add `affiliate` and `affiliates` methods that raise NotImplementedError, to ensure that child classes implement them.

dmann commented on app/models/ldap_identity_provider.rb:
```diff
@@ -0,0 +1,37 @@
+    )
+  end
+
+  def ldap_search(ldap_filter)
```
I think `ldap_search` should be private.

dmann commented on spec/factories/users.rb:
```diff
@@ -1,7 +1,7 @@
   factory :user do
     etag { SecureRandom.hex }
-    sequence(:username) { |n| "#{Faker::Internet.user_name}_#{n}" }
+    sequence(:username) { |n| "#{Faker::Name.first_name}_#{n}" }
```
`Faker::Name.first_name` includes names with apostrophes, like D'Angelo, which can lead to transient test failures.  :(

dmann commented on spec/models/ldap_identity_provider_spec.rb:
```diff
@@ -0,0 +1,78 @@
+  end
+
+  describe 'affiliate' do
+    it { is_expected.to respond_to(:affiliate) }
```
You can get rid of the following context by replacing the above with:
```
it { is_expected.to respond_to(:affiliate).with(1).argument }
```

dmann replied:
https://www.relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/respond-to-matcher

dmann commented on spec/models/ldap_identity_provider_spec.rb:
```diff
@@ -0,0 +1,78 @@
+    it { is_expected.to validate_presence_of :ldap_base }
+  end
+
+  describe 'affiliate' do
```
Nit picky, but please prepend method names with `#` for instance and `::` for class methods.

dmann commented on spec/models/ldap_identity_provider_spec.rb:
```diff
@@ -0,0 +1,78 @@
+    end
+  end
+
+  describe 'affiliates' do
```
Please prepend method name with `#`

dmann commented on spec/support/ldap_shared_examples.rb:
```diff
@@ -0,0 +1,91 @@
+    }
+  }
+  before do
+    allow_any_instance_of(Net::LDAP).to receive(:search).and_return(expected_entries)
```
Cool! I'm sure there are other places in our code that could use this technique.

dmann reviewed (changes requested):
Looks good for the most part. Just asking for a few minor changes.

dmlond committed:
uses standard for naming instance metthod tests, implements interface methods that raise NotImplemented

dmlond committed:
ldap_search is private
asserts that LdapIdentityProvider is_an IdentityProvider
no longer needs to test that it throws argument errors when calling affiliate without an argument
use standard naming convention for tests of instance methods

dmlond committed:
use Faker::Internet.user_name instead of Faker::Name.first_name

dmlond committed:
Merge pull request #907 from dmlond/DDS-715_user_factory_fix

Dds 715 user factory fix

dmlond commented:
all but pagination are addressed

dmlond committed:
Merge branch 'develop' into DDS-715_auth_service_affiliates

dmann committed:
Mock Net::LDAP.search to return Net::LDAP::Error

dmann committed:
Move identified resource tests into their own contexts for better readability

dmann committed:
Refactor of the list auth_provider affiliates endpoint

dmann committed:
Use shared example 'a GET request' in preparation for pagination requirement

dmann committed:
Paginate the List auth_providers affiliates endpoint

dmann committed:
Merge tag 'v1.1.2' into DDS-715_auth_service_affiliates

Fix request timeout using Rack::Timeout gem

dmann committed:
Merge tag 'v1.1.2' into DDS-715_auth_service_affiliates

Fix request timeout using Rack::Timeout gem

dmann committed:
Merge pull request #919 from dmann/DDS-715_auth_service_affiliates

Merging in v1.1.2

dmann committed:
Merge pull request #920 from dmann/DDS-807-list_affiliates_pagination

DDS-807 list affiliates pagination
